File: frontend/src/app/services/sites/qwirkle.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Globals} from "../../classes/globals";
import {Observable} from "rxjs";
import {GameState} from "../../dto/sites/qwirkle/game/GameState";
import {Move} from "../../dto/sites/qwirkle/move/Move";
import {Tile} from "../../dto/sites/qwirkle/tile/Tile";
import {MoveGroup} from "../../dto/sites/qwirkle/move/MoveGroup";
import {SelectionInfo} from "../../dto/sites/qwirkle/SelectionInfo";
import {GameSessionService} from "./game-session.service";
import {GameConfig} from "../../dto/sites/qwirkle/game/GameConfig";
import {State} from "@popperjs/core";

@Injectable({
    providedIn: 'root'
})
export class QwirkleService extends GameSessionService<GameConfig, GameState> {
    protected readonly baseUri: string;

    constructor(httpClient: HttpClient, globals: Globals) {
        super(httpClient, globals);
        this.baseUri = this.globals.backendUri + '/qwirkle';
    }

    clearHand(): Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/hand/clear', {});
    }

    getSelectionInfo(selected: Tile[], fromStack: boolean = false): Observable<SelectionInfo> {
        return this.httpClient.post<SelectionInfo>(this.baseUri + '/selection', selected, {
            params: {
                fromStack: fromStack
            }
        });
    }

    getBestMoves(maxMoves: number = 1): Observable<MoveGroup[]> {
        return this.httpClient.get<MoveGroup[]>(this.baseUri + '/solve', {
            params: {maxMoves: maxMoves.toString()}
        });
    }

    drawTile(tile: Tile): Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/stack/draw', tile);
    }

    makeMove(move: Move, fromStack: boolean = false): Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/move', move, {
            params: {
                fromStack: fromStack
            }
        });
    }

    undoMove() : Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/undo', {});
    }

    uploadImage(image: File): Observable<Blob> {
        const formData = new FormData();
        formData.append('file', image);

        return this.httpClient.post<Blob>(this.baseUri + '/image/upload', formData, {
            responseType: 'blob' as 'json'
        }) as Observable<Blob>;
    }

    confirmImage(): Observable<void> {
        return this.httpClient.post<void>(this.baseUri + '/image/confirm', {});
    }
}

File: backend/src/main/java/net/zorphy/backend/site/qwirkle/service/QwirkleService.java
package net.zorphy.backend.site.qwirkle.service;

import net.zorphy.backend.site.GameSessionService;
import net.zorphy.backend.site.qwirkle.dto.game.GameConfig;
import net.zorphy.backend.site.qwirkle.dto.game.GameState;
import net.zorphy.backend.site.qwirkle.dto.SelectionInfo;
import net.zorphy.backend.site.qwirkle.dto.move.Move;
import net.zorphy.backend.site.qwirkle.dto.move.MoveGroup;
import net.zorphy.backend.site.qwirkle.dto.tile.Tile;

import java.util.List;

public interface QwirkleService extends GameSessionService<GameConfig, GameState> {
    /**
     * Retrieves the {@code N} best moves for the given {@code gameState} where {@code N} is {@code maxMoves}.
     */
    List<MoveGroup> getBestMoves(GameState gameState, int maxMoves);

    /**
     * Draws the given {@code tile} from the stack in the given {@code gameState}.
     */
    GameState drawTile(GameState gameState, Tile tile);

    /**
     * Retrieves the selection information for the given {@code tiles} in the given {@code gameState}.
     */
    SelectionInfo getSelectionInfo(GameState gameState, List<Tile> tiles, boolean fromStack);

    /**
     * Clears the hand in the given {@code gameState}.
     */
    GameState clearHand(GameState gameState);

    /**
     * Makes the given {@code move} in the given {@code gameState}.
     */
    GameState makeMove(GameState gameState, Move move, boolean fromStack);

    byte[] uploadImage(byte[] file);
}
